### PR TITLE
Improvment to `Array.isArray`

### PR DESCRIPTION
**Pull request by Mickemouse0**
This PR improves the type inference of `Array.isArray`.

It builds on the changes made in https://github.com/total-typescript/ts-reset/pull/56 and https://github.com/total-typescript/ts-reset/pull/23 to solve https://github.com/total-typescript/ts-reset/issues/48.

### CODE REVIEW

**Comment by Mickemouse0** [comment]
@mattpocock @DeepDoge @phenomnomnominal I have not been able to make it pass the final test. If you have any suggestions I'd be happy for the help.

**Comment by Mickemouse0** [comment]
I have been able to get either or of
```typescript
doNotExecute(() => {
  function test<T>(value: T) {
    type Unarray<T> = T extends Array<infer U> ? U : T;
    const inner = <X extends Unarray<T>>(v: X[]) => {};

    if (Array.isArray(value)) {
      inner(value);
    }
  }
});

doNotExecute(async () => {
  function makeArray<Item>(input: Item | ReadonlyArray<Item> | Array<Item>) {
    if (Array.isArray(input)) {
      return input;
    }
    return [input];
  }

  const [first] = makeArray([{ a: "1" }, { a: "2" }, { a: "3" }] as const);

  // No error!
  first.a;
});
``` 
to pass but not both at the same time.

With
```typescript
interface ArrayConstructor {
  isArray<T>(arg: true extends TSReset.IsAny<T> ? never : T): arg is true extends TSReset.IsAny<T> ? never : T extends unknown[] | readonly unknown[] ? T : T[] extends T ? T[] : never;
}
``` 
the first passes and with
```typescript
interface ArrayConstructor {
  isArray<T>(arg: true extends TSReset.IsAny<T> ? never : T): arg is true extends TSReset.IsAny<T> ? never : T extends unknown[] | readonly unknown[] ? T : T[] extends T ? T[] : never;
  isArray<T>(arg: T[] extends T ? T | T[] : never): arg is T[] extends T ? T[] : never;
  isArray<T>(arg: T): arg is T extends unknown[] | readonly unknown[] ? T : T[] extends T ? T[] : never;
}
``` 
the second passes.

Any help to get both to pass would be greatly appreciated.

**Comment by DeepDoge** [comment]
```ts
doNotExecute(async () => {
  function makeArray<Item>(input: Item) {
    if (Array.isArray(input)) {
      return input;
    }
    return [input];
  }

  const [first] = makeArray([{ a: "1" }, { a: "2" }, { a: "3" }] as const);

  // No error!
  first.a;
});
```

In the test above, you get a value, and if that value is an array, you just return it, if its not an array you place it inside an array and return it. 

So as far as typescript knows that `makeArray` function can either return `[Item]` or `Item`, so it's return type is `Item | [Item]`. 
It doesn't care about your runtime logic. 
So this is a typescript issue, how it works, can't fix it.

Normally to make it give the right type you would define the `makeArray` function like this.
```ts
function makeArray<Item>(input: Item): 
    Item extends Array<any> ? Item : 
    Item extends ReadonlyArray<any> ? Item : 
    [Item]
  {
     if (Array.isArray(input)) {
       return input as any;
     }
     return [input] as any;
  }
```
So, basically you have to repeat your runtime logic with types. It's a typescript issue, not an issue with `isArray()`.

As far as I know, but a second opinion won't hurt, maybe I'm missing something. 
Tbh, I was thinking about making my own scripting or programming language to overcome this issue and more.

**Comment by Mickemouse0** [comment]
Thanks @DeepDoge for your answer. If @mattpocock agrees I could just remove that test and it would be ready from my side.

File: src/entrypoints/is-array.d.ts [hunk 1]
```diff
@@ -1,3 +1,3 @@
 interface ArrayConstructor {
-  isArray(arg: any): arg is unknown[];
+  isArray<T>(arg: 0 extends 1 & T ? never : T): arg is 0 extends 1 & T ? never : T extends unknown[] | readonly unknown[] ? T : T[] extends T ? T[] : never;
```
Comment (ahrjarrett) [inline src/entrypoints/is-array.d.ts:2]: Not sure if you saw my comment [here](https://github.com/total-typescript/ts-reset/pull/56#issuecomment-1595405064) -- you can simplify this by removing the check for `unknown[]` entirely.

```suggestion
  isArray<T>(arg: 0 extends 1 & T ? never : T): arg is 0 extends 1 & T ? never : T extends readonly unknown[] ? T : T[] extends T ? T[] : never;
```

Comment (Mickemouse0) [reply src/entrypoints/is-array.d.ts]: @ahrjarrett Thanks, I saw your comment but it slipped my mind while I was sitting with the code

**Comment by Maxim-Mazurok** [comment]
I've tried [this](https://github.com/total-typescript/ts-reset/pull/56#issuecomment-1595417081) and this PR, none of them work for my case where I want to narrow type from string or string array:
```ts
const token: string | readonly string[];
if (Array.isArray(token)) throw new Error("Token is an array");
console.log(token); // token is `string | readonly string[]` here; expected just `string`
```

**Comment by Mickemouse0** [comment]
@mattpocock I removed the failing test based on [this comment](https://github.com/total-typescript/ts-reset/pull/151#issuecomment-1596060662). This PR is now redo to be merged

File: src/entrypoints/is-array.d.ts [hunk 2]
```diff
@@ -1,3 +1,3 @@
 interface ArrayConstructor {
-  isArray(arg: any): arg is unknown[];
+  isArray<T>(arg: true extends TSReset.IsAny<T> ? never : T): arg is true extends TSReset.IsAny<T> ? never : T extends ReadonlyArray<unknown> ? T : Array<T> extends T ? Array<T> : never;
 }
```
Comment (ahrjarrett) [inline src/entrypoints/is-array.d.ts:3]: Not sure I agree with the decision to disallow `any`, since narrowing a value from `any` to `unknown[]` seems like a valid thing to want to do.

Setting that aside, here's another way you could accomplish the same thing:

```suggestion
interface ArrayConstructor {
  isArray<T>(arg: NotAny<T>): arg is Extract<NotAny<T>, readonly unknown[]>;
}

type NotAny<T> = 0 extends 1 & T ? never : T
```

**Comment by Mickemouse0** [comment]
@mattpocock Are the changes in this PR something that will be considered?
